File: src/peated/models/entity.py
from geoalchemy2 import Geometry
from sqlalchemy import (
    ARRAY,
    BigInteger,
    Column,
    DateTime,
    Enum,
    ForeignKeyConstraint,
    Index,
    PrimaryKeyConstraint,
    Text,
    text,
)

from peated.db.base import Base


class Entity(Base):
    __tablename__ = "entity"
    __table_args__ = (
        ForeignKeyConstraint(["created_by_id"], ["user.id"], name="entity_created_by_id_user_id_fk"),
        PrimaryKeyConstraint("id", name="entity_pkey"),
        Index("entity_name_unq", "name", unique=True),
    )

    id = Column(BigInteger)
    name = Column(Text, nullable=False)
    type = Column(
        ARRAY(
            Enum(
                "brand",
                "distiller",
                "bottler",
                name="entity_type",
                _create_events=False,
            )
        ),
        nullable=False,
    )
    created_at = Column(DateTime, nullable=False, server_default=text("now()"))
    created_by_id = Column(BigInteger, nullable=False)
    total_bottles = Column(BigInteger, nullable=False, server_default=text("0"))
    total_tastings = Column(BigInteger, nullable=False, server_default=text("0"))
    country = Column(Text)
    region = Column(Text)
    location = Column(Geometry(geometry_type="POINT", srid=4326))

File: src/server/db/model/entity.go
package model

import (
	"time"

	"gorm.io/gorm"
)

type EntityType string

const (
	EntityTypeBottler   EntityType = "bottler"
	EntityTypeBrand     EntityType = "brand"
	EntityTypeDistiller EntityType = "distiller"
)

type Entity struct {
	gorm.Model
	ID        uint64       `gorm:"primaryKey" json:"id"`
	Name      string       `json:"name"`
	ShortName string       `json:"short_name"`
	Country   string       `json:"country"`
	Region    string       `json:"region"`
	Type      []EntityType `json:"type"`

	Description     string `json:"description"`
	YearEstablished string `json:"year_established"`
	Website         string `json:"website"`

	TotalBottles  uint `json:"total_bottles"`
	TotalTastings uint `json:"total_tastings"`

	CreatedAt   time.Time `json:"created_at"`
	CreatedByID uint64    `json:"created_by_id"`

	CreatedBy User
}

func (Entity) TableName() string {
	return "entity"
}

type Entities []*Entity
